File: 3548.equal-sum-grid-partition-ii.py
#
# @lc app=leetcode id=3548 lang=python3
#
# [3548] Equal Sum Grid Partition II
#

from typing import List, Dict

# @lc code=start
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])

        # Precompute sums and prepare column lists for vertical sweep.
        rowSum = [0] * m
        colSum = [0] * n
        cols = [[] for _ in range(n)]

        total = 0
        overall: Dict[int, int] = {}
        for i in range(m):
            rsum = 0
            row = grid[i]
            for j, v in enumerate(row):
                rsum += v
                colSum[j] += v
                cols[j].append(v)
                total += v
                overall[v] = overall.get(v, 0) + 1
            rowSum[i] = rsum

        def inc(mp: Dict[int, int], x: int, d: int) -> None:
            nv = mp.get(x, 0) + d
            if nv:
                mp[x] = nv
            else:
                mp.pop(x, None)

        def eligible(delta: int, mp: Dict[int, int], r0: int, r1: int, c0: int, c1: int) -> bool:
            h = r1 - r0 + 1
            w = c1 - c0 + 1
            if h * w <= 1:
                return False
            if h == 1 or w == 1:
                # Only endpoints removable in a 1D path.
                if r0 == r1:
                    return delta == grid[r0][c0] or delta == grid[r0][c1]
                else:
                    return delta == grid[r0][c0] or delta == grid[r1][c0]
            # 2D rectangle: any cell removable.
            return mp.get(delta, 0) > 0

        # Horizontal sweep
        top: Dict[int, int] = {}
        bottom: Dict[int, int] = dict(overall)
        topSum = 0

        for r in range(m - 1):  # cut after row r
            for v in grid[r]:
                inc(top, v, 1)
                inc(bottom, v, -1)
            topSum += rowSum[r]
            botSum = total - topSum

            if topSum == botSum:
                return True
            if topSum > botSum:
                delta = topSum - botSum
                if eligible(delta, top, 0, r, 0, n - 1):
                    return True
            else:
                delta = botSum - topSum
                if eligible(delta, bottom, r + 1, m - 1, 0, n - 1):
                    return True

        # Vertical sweep
        left: Dict[int, int] = {}
        right: Dict[int, int] = dict(overall)
        leftSum = 0

        for c in range(n - 1):  # cut after col c
            for v in cols[c]:
                inc(left, v, 1)
                inc(right, v, -1)
            leftSum += colSum[c]
            rightSum = total - leftSum

            if leftSum == rightSum:
                return True
            if leftSum > rightSum:
                delta = leftSum - rightSum
                if eligible(delta, left, 0, m - 1, 0, c):
                    return True
            else:
                delta = rightSum - leftSum
                if eligible(delta, right, 0, m - 1, c + 1, n - 1):
                    return True

        return False
# @lc code=end

File: 3548.equal-sum-grid-partition-ii.cpp
#
# @lc app=leetcode id=3548 lang=cpp
#
# [3548] Equal Sum Grid Partition II
#
# @lc code=start
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        for (int i = 0; i < m - 1; i++) {
            if (checkHorizontalCut(grid, i)) return true;
        }
        
        for (int j = 0; j < n - 1; j++) {
            if (checkVerticalCut(grid, j)) return true;
        }
        
        return false;
    }
    
private:
    bool checkHorizontalCut(vector<vector<int>>& grid, int cutRow) {
        int m = grid.size(), n = grid[0].size();
        long long topSum = 0, bottomSum = 0;
        
        for (int i = 0; i <= cutRow; i++)
            for (int j = 0; j < n; j++)
                topSum += grid[i][j];
        
        for (int i = cutRow + 1; i < m; i++)
            for (int j = 0; j < n; j++)
                bottomSum += grid[i][j];
        
        if (topSum == bottomSum) return true;
        
        if (topSum > bottomSum) {
            long long diff = topSum - bottomSum;
            for (int i = 0; i <= cutRow; i++)
                for (int j = 0; j < n; j++)
                    if (grid[i][j] == diff && isConnected(grid, 0, cutRow, 0, n - 1, i, j))
                        return true;
        }
        
        if (bottomSum > topSum) {
            long long diff = bottomSum - topSum;
            for (int i = cutRow + 1; i < m; i++)
                for (int j = 0; j < n; j++)
                    if (grid[i][j] == diff && isConnected(grid, cutRow + 1, m - 1, 0, n - 1, i, j))
                        return true;
        }
        
        return false;
    }
    
    bool checkVerticalCut(vector<vector<int>>& grid, int cutCol) {
        int m = grid.size(), n = grid[0].size();
        long long leftSum = 0, rightSum = 0;
        
        for (int i = 0; i < m; i++)
            for (int j = 0; j <= cutCol; j++)
                leftSum += grid[i][j];
        
        for (int i = 0; i < m; i++)
            for (int j = cutCol + 1; j < n; j++)
                rightSum += grid[i][j];
        
        if (leftSum == rightSum) return true;
        
        if (leftSum > rightSum) {
            long long diff = leftSum - rightSum;
            for (int i = 0; i < m; i++)
                for (int j = 0; j <= cutCol; j++)
                    if (grid[i][j] == diff && isConnected(grid, 0, m - 1, 0, cutCol, i, j))
                        return true;
        }
        
        if (rightSum > leftSum) {
            long long diff = rightSum - leftSum;
            for (int i = 0; i < m; i++)
                for (int j = cutCol + 1; j < n; j++)
                    if (grid[i][j] == diff && isConnected(grid, 0, m - 1, cutCol + 1, n - 1, i, j))
                        return true;
        }
        
        return false;
    }
    
    bool isConnected(vector<vector<int>>& grid, int r1, int r2, int c1, int c2, int skipR, int skipC) {
        int totalCells = (r2 - r1 + 1) * (c2 - c1 + 1) - 1;
        if (totalCells <= 0) return false;
        
        int startR = -1, startC = -1;
        for (int i = r1; i <= r2 && startR == -1; i++)
            for (int j = c1; j <= c2; j++)
                if (i != skipR || j != skipC) {
                    startR = i; startC = j; break;
                }
        
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        queue<pair<int, int>> q;
        q.push({startR, startC});
        visited[startR][startC] = true;
        int count = 1;
        
        int dx[] = {-1, 1, 0, 0}, dy[] = {0, 0, -1, 1};
        
        while (!q.empty()) {
            auto [x, y] = q.front(); q.pop();
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d], ny = y + dy[d];
                if (nx >= r1 && nx <= r2 && ny >= c1 && ny <= c2 && 
                    !visited[nx][ny] && (nx != skipR || ny != skipC)) {
                    visited[nx][ny] = true;
                    q.push({nx, ny});
                    count++;
                }
            }
        }
        
        return count == totalCells;
    }
};
# @lc code=end